### PR TITLE
PARQUET-1571: [C++] Can't read data from parquet file in C++ library

### PR DESCRIPTION
**Pull request by worker24h**
JIRA issue: https://issues.apache.org/jira/browse/PARQUET-1571
Github Issues:  Fixed Issues #4256

### CODE REVIEW

**Comment by kou** [comment]
Could you open a JIRA issue at https://issues.apache.org/jira/browse/ARROW and use "ARROW-${ISSUE_NUMBER}: [C++] ${SUMMARY}" format as the pull request title?
See also: https://arrow.apache.org/docs/developers/contributing.html#how-to-contribute-patches

**Comment by wesm** [comment]
@worker24h can you write a unit test that exhibits the problem?

**Comment by worker24h** [comment]
> Could you open a JIRA issue at https://issues.apache.org/jira/browse/ARROW and use "ARROW-${ISSUE_NUMBER}: [C++] ${SUMMARY}" format as the pull request title?
> See also: https://arrow.apache.org/docs/developers/contributing.html#how-to-contribute-patches

I had changed

**Comment by worker24h** [comment]
> @worker24h can you write a unit test that exhibits the problem?
OK, i'm writing now....

File: cpp/src/parquet/util/memory.cc [hunk 1]
```diff
@@ -270,7 +270,7 @@ const uint8_t* BufferedInputStream::Peek(int64_t num_to_peek, int64_t* num_bytes
     DCHECK(buffer_size_ >= *num_bytes);
   }
   // Read more data when buffer has insufficient left or when resized
-  if (*num_bytes > (buffer_size_ - buffer_offset_)) {
+  if (*num_bytes >= (buffer_size_ - buffer_offset_)) {
```
Comment (pitrou) [inline cpp/src/parquet/util/memory.cc]: I don't think that's the problem. You only need to read more data if you want strictly more than what's available in the buffer (so `>`).

There are other problems though:
- there is a confusion with `buffer_size_`, which claims to track the number of bytes available, but we change it in `Resize` above without reading more bytes; really, the code should distinguish between buffer size and capacity
- `Peek()` ignores any pending data in the buffer if `num_to_peek` is larger than the buffer size

**Comment by pitrou** [comment]
I submitted an alternate PR (+ tests) at #4275.

**Comment by pitrou** [comment]
Closing as #4275 was merged. @worker24h can you check your use case works with git master?

**Pull request closed without merging**